### PR TITLE
added get_research_authors method to ResearchDetailPage

### PR DESCRIPTION
**Pull request by danielfmiranda**
# Description

Related PRs/issues: #10418 

**This PR adds the following:**
- A new method for the ResearchDetailPage model called `get_research_authors`. This method returns a localized queryset of all the `Profiles`  for research authors that are related to  the page through a [`ResearchAuthorRelation`](https://github.com/MozillaFoundation/foundation.mozilla.org/blob/main/network-api/networkapi/wagtailpages/pagemodels/research_hub/relations.py#L7). I then call this method to add the list of profiles to the page's template context for use.
- Update to the template `research_detail_page.html` to now use the profile information from the context, instead of `page.research_authors.all` and then `research_author.author_profile`
- Some tests for the method to make sure it works and returns localized profiles as expected

### CODE REVIEW

File: network-api/networkapi/wagtailpages/pagemodels/research_hub/detail_page.py
Comment (jhonatan-lopes) [inline network-api/networkapi/wagtailpages/pagemodels/research_hub/detail_page.py:144]: I think we need a `prefetch_related` here to avoid the N+1 problem, something like:

```python
Profile.objects.prefetch_related("authored_research").filter(authored_research__research_detail_page=self)
```
Otherwise, for each `authored_research` in a `Profile`, we will have additional queries.

File: network-api/networkapi/wagtailpages/tests/research_hub/test_detail_page.py
Comment (jhonatan-lopes) [inline network-api/networkapi/wagtailpages/tests/research_hub/test_detail_page.py:39]: Here, it would be interesting to add a `Profile` linking to another page, and check that it doesn't show up in our `research_authors`.

File: network-api/networkapi/wagtailpages/tests/research_hub/test_detail_page.py
Comment (jhonatan-lopes) [inline network-api/networkapi/wagtailpages/tests/research_hub/test_detail_page.py:55]: I think it would be best to assert that the author profiles actually show up in `research_authors` instead of counts, something like:

```python
self.assertIn(author_profiles[0], research_authors)
self.assertIn(author_profiles[1], research_authors)
self.assertIn(author_profiles[2], research_authors)
```

And if we had a profile that was not supposed to be in `research_authors`:

```python
self.assertNotIn(profile_not_in_this_page, research_authors)
```

File: network-api/networkapi/wagtailpages/tests/research_hub/test_detail_page.py
Comment (jhonatan-lopes) [inline network-api/networkapi/wagtailpages/tests/research_hub/test_detail_page.py:52]: Can we [assert the number of queries](https://docs.djangoproject.com/en/4.2/topics/testing/tools/#django.test.TransactionTestCase.assertNumQueries) here, please? We need to make sure that we are not running into a N+1 problem again.

**Review by jhonatan-lopes — CHANGES_REQUESTED** [review]
Thanks for implementing this @danielfmiranda! Could you take a look at the comments here, please?

File: network-api/networkapi/wagtailpages/pagemodels/research_hub/detail_page.py
Comment (danielfmiranda) [reply network-api/networkapi/wagtailpages/pagemodels/research_hub/detail_page.py:144]: Hey Jhonatan, thanks for walking through this comment with me over zoom. I have updated the PR to now include the `prefetch_related` method!

File: network-api/networkapi/wagtailpages/tests/research_hub/test_detail_page.py
Comment (danielfmiranda) [reply network-api/networkapi/wagtailpages/tests/research_hub/test_detail_page.py:55]: Thanks for the suggestion Jhonatan!

[AssertCountEqual](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertCountEqual) actually does not check the count of items, but instead checks that two items contain the same elements, regardless of their order. 

However, after your suggestion I can see that the name is very misleading, so much so that there has been a [github issue opened over it](https://github.com/python/cpython/issues/71258)! Though in the ticket they decided to leave the name as-is.

I like your suggestion of checking for each item with their index instead, as its easier to read/understand and have added it to my PR 👍 

I also added the other suggestion of creating an unrelated page and author and double checking that they do not appear in the list.

File: network-api/networkapi/wagtailpages/tests/research_hub/test_detail_page.py
Comment (danielfmiranda) [reply network-api/networkapi/wagtailpages/tests/research_hub/test_detail_page.py:39]: Updated! See comment below

File: network-api/networkapi/wagtailpages/tests/research_hub/test_detail_page.py
Comment (danielfmiranda) [reply network-api/networkapi/wagtailpages/tests/research_hub/test_detail_page.py]: Hi @jhonatan-lopes! 

Thanks for working through this with me over zoom, instead of doing the check here, I have added a new function titled "[test_get_research_authors_avoids_n1_queries](https://github.com/MozillaFoundation/foundation.mozilla.org/blob/16637e81be7af07e285942f16607e9e59ef428b2/network-api/networkapi/wagtailpages/tests/research_hub/test_detail_page.py#L63)" as discussed.

It creates 10 related authors, and then checks that the number of queries are less than that.

One concern I do have however and one I was hoping you might be able to help with, is when taking a look at the queries that the assert statement catches (currently catches 2), the queries both seem to be locale related 🤔 


Screenshot 2023-04-26 at 5 39 27 PM

From my understanding, I believe there should be at least 4:
```python
research_author_profiles = localize_queryset(
   Profile.objects.prefetch_related("authored_research").filter(authored_research__research_detail_page=self)
)
```
- 1 for pulling the profiles
- 1 for the `prefetch` statement from the PR feedback
- 1 for `localize_queryset` to get the default locale
- 1 for `localize_queryset` to get the active locale 

I do remember us discussing that the number of queries can vary from time to time. 

Should I wait until we merge in the testing framework from #10536 so I can change the statement to something along the lines of `self.assertNumQueries(< 10)`?

**Comment by danielfmiranda** [comment]
Hi @jhonatan-lopes, thanks for the review! I have updated this PR based on your feedback and this PR is now ready for re-review whenever you find a free moment

File: network-api/networkapi/wagtailpages/tests/research_hub/test_detail_page.py
Comment (jhonatan-lopes) [reply network-api/networkapi/wagtailpages/tests/research_hub/test_detail_page.py]: Using the debug toolbar on a detail page, I can see that there is a query on profile: 
image

We know the prefetch is working since we can change the N+1 test to run with 100 or 1000 objects and it takes (roughly) the same amount of time. I won't block the PR because of this issue. If you would like to investigate this further, it's up to you. Otherwise, we can remove this test and create a ticket to revisit this another time.

**Review by jhonatan-lopes — APPROVED** [review]
Thank you for the changes, Daniel! I've put a comment for the number of queries assertion but it's non-blocking. LGTM :+1:

File: network-api/networkapi/wagtailpages/tests/research_hub/test_detail_page.py
Comment (danielfmiranda) [reply network-api/networkapi/wagtailpages/tests/research_hub/test_detail_page.py]: Thanks @jhonatan-lopes! 

I have created https://github.com/MozillaFoundation/foundation.mozilla.org/issues/10554 to track that work 👍

**Pull request merged**
